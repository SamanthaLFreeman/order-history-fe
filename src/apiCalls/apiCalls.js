
// export const fetchOrders = () => {
//   fetch('http://localhost:3001/api/v1/purchases')
//     .then(response => {
//       if(!response.ok) {
//         throw Error('Error fetching orders')
//       }
//       return response.json()
//     })
//     .catch(error => {
//       throw Error(error.message)
//     })
// }
