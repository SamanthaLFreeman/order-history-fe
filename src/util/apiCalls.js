
export const fetchOrders = () => {
  return fetch('http://localhost:3001/api/v1/purchases')
    .then(response => {
      if(!response.ok) {
        throw Error('Error fetching orders')
      }
      return response.json()
    })
    .catch(error => {
      throw Error(error.message)
    })
}

export const postOrder = (newOrder) => {
  const options = {
    method: 'POST',
    body: JSON.stringify(newOrder),
    headers: {
      'Content-Type': 'application/json'
    }
  }

  return fetch('http://localhost:3001/api/v1/purchases', options)
    .then(response => {
      if (!response.ok) {
        throw Error('Error posting orders')
      }
      return response.json()
    })
    .catch(error => {
      throw Error(error.message)
    })
}

export const deleteOrder = (id) => {
  const options = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    }
  }

  return fetch(`http://localhost:3001/api/v1/purchases/${id}`, options)
    .then(() => fetch('http://localhost:3001/api/v1/purchases'))
    .then(response => {
      if(!response.ok) {
        throw Error('Error deleting orders')
      }
      return response.json()
    })
    .catch(error => {
      throw Error(error.message)
    })
}