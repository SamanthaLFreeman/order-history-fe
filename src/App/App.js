import React, { Component } from 'react';
import './App.css';
import OrdersContainer from '../OrdersContainer/OrdersContainer';
import OrdersForm from '../OrdersForm/OrdersForm';
import { fetchOrders, postOrder, deleteOrder } from '../util/apiCalls';

class App extends Component {
  constructor() {
    super();
    this.state = {
      orders: []
    }
  }

  componentDidMount() {
    fetchOrders()
      .then(data => this.setState({orders: data}))
      .catch(error => console.log(error))
  }

  addOrder = (newOrder) => {
    postOrder(newOrder)
      .then(() => this.setState({ orders: [...this.state.orders, newOrder] }))
      .catch(error => console.log(error));
  }

  removeOrder = (id) => {
    const filteredOrders = this.state.orders.filter(order => {
      return order.id !== id
    })  

    deleteOrder(id)
      .then(() => this.setState({orders: filteredOrders}))
      .catch(error => console.log(error))
  }

  render() {
    return (
      <div className="App">
        <header>
          <h1 className='app-title'>My Order History</h1>
          <div className='purchase-form'>
            <OrdersForm addOrder={this.addOrder}/>
          </div>
        </header>
        <div className='purchase-container'>
          <OrdersContainer orders={this.state.orders} removeOrder={this.removeOrder}/>
        </div>
      </div>
    );
  }
}

export default App;
