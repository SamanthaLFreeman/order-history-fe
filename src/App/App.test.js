import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import { fetchOrders, postOrder, deleteOrder } from '../util/apiCalls'

jest.mock('../util/apiCalls.js');


describe('App', () => {

  beforeEach(() => {

    fetchOrders.mockImplementation(() => {
      return Promise.resolve([{
        id: 42,
        name: "Blah",
        img: "Blah img",
        description: "Blah blah blah",
        price: "1000000"
      }])
    })

    postOrder.mockImplementation(() => {
      return Promise.resolve([{
        id: 43,
        name: "Blah",
        img: "Blah img",
        description: "Blah blah blah",
        price: "1000000"
      }] )
    })

    deleteOrder.mockImplementation(() => {
      return Promise.resolve([])
    })
  })

  it('should match the snapshot', async () => {
    let wrapper = shallow(<App />)
    expect(wrapper).toMatchSnapshot();
  })

  it('should call fetchOrders after mounting', () => {
    shallow(<App />);
    expect(fetchOrders).toHaveBeenCalled();
  })

  it('should update state with an order when addOrder is called', async () => {
    // Setup
    const wrapper = await shallow(<App />);
    const mockOrder = {
      id: 43,
      name: "Blah",
      img: "Blah img",
      description: "Blah blah blah",
      price: "1000000"
    };
    const expected = [{
      id: 42,
      name: "Blah",
      img: "Blah img",
      description: "Blah blah blah",
      price: "1000000"
    }, mockOrder];

    // Execution
    await wrapper.instance().addOrder(mockOrder);

    // Expectation
    expect(wrapper.state('orders')).toEqual(expected);
  });

  it('should update state with the order removed when removeOrder is called', async () => {
    // Setup
    const wrapper = await shallow(<App />);
    const mockOrder = [{
      id: 42,
      name: "Blah",
      img: "Blah img",
      description: "Blah blah blah",
      price: "1000000"
    }];
    const expected = [];

    expect(wrapper.state('orders')).toEqual(mockOrder)
    // Execution
    await wrapper.instance().removeOrder(42);

    // Expectation
    expect(wrapper.state('orders')).toEqual(expected);
  })
})
