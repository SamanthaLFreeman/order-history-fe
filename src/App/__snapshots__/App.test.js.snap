// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="App"
>
  <header>
    <h1
      className="app-title"
    >
      My Order History
    </h1>
    <div
      className="purchase-form"
    >
      <OrdersForm
        addOrder={[Function]}
      />
    </div>
  </header>
  <div
    className="purchase-container"
  >
    <OrdersContainer
      orders={Array []}
      removeOrder={[Function]}
    />
  </div>
</div>
`;
